proposals/0000-wire-protocol.md

```diff
@@ -0,0 +1,485 @@
+
+Authors: [Paul Frazee](https://github.com/pfrazee), [Bryan Newbold](https://github.com/bnewbold)
+
+
+# Summary
+[summary]: #summary
+
+This DEP describes the Hypercore wire protocol: a transport-agnostic message stream spoken between nodes in a swarm of hypercore network peers (including Dat clients). The wire protocol includes mechanisms for framing, stream encryption, and feed key authentication.
+
+
+# Motivation
+[motivation]: #motivation
+
+The protocol described here is already in use as of 2017 (by Hypercore, Dat, and Beaker Browser users), and was partially described in an earlier [whitepaper][whitepaper]. This document fills in some additional details.
+
+[whitepaper]: https://github.com/datprotocol/whitepaper
+
+
+# Stream Connections
+[stream-details]: #stream-details
+
+The Dat wire protocol depends on the use of a binary transport protocol which provides the following semantics:
+
+- reliable delivery (no dropped or partial messages)
+- in-order delivery of messages
+
+Two notable transport protocols that satisfy these requirements are TCP and µTP (the "Micro Transport Protocol").
+
+Peers wishing to connect need to discover each other using some mechanism or another (see forthcoming DEPs on some options; this process is modular and swappable), and both need to have the public key for the primary hypercore they wish to exchange.
+
+Messages are framed by the Dat protocol itself (see Messages section for details).
+
+
+## Channels
+[channels]: #channels
+
+Multiple hypercore feeds can be synchronized over the same protocol connection. Messages pertaining to the separate channels are tagged with an id for disambiguation.
+
+Note that at least one feed is necessary for each connection (for handshaking to succeed), and that the first feed is the one used for discovery and as an encryption key.
+
+To initiate a new channel (after the primary is established), a peer can send a Feed message, followed by an Info message. Unlike the first message sent on an overall connection, later Feed messages are encrypted. Either party may initiate a new channel with a Feed message at any time.
+
+
+## Handshake Procedure
+[handshake]: #handshake
+
+A handshake procedure needs to occur for each feed on a channel; the first part of the first handshake happens in cleartext and both validates discovery keys and establishes encyption parameters used for the rest of the connection. The first (primary) channel has `id=0`.
+
+The first (cleartext) message is a Feed message, and includes two fields: a nonce and a discovery key.
+
+The **nonce** is generated by each peer as a random 32-byte sequence.
+
+The **discovery key** is generated from the public encryption key for a hypercore feed (in this case the first, or "primary" feed) by using the public key to sign the 9-byte ASCII string "HYPERCORE" (with no trailing NULL byte) using the BLAKE2b keyed hashing algorithm (provided by most BLAKE2b hash implementations). The discovery key is 32 bytes long.
```

Comment:
It is not actually signing anything, but performing a keyed hash (as correctly described). It is a core part of the BLAKE2b algorithm, and is also called a Prefix MAC, since you essentially do `blake2b(HYPERCORE || publicKey)`. Signing would in my view involve the private key, which only the writer has access to.